[rbd] fix rbd help for image-name and snap-name

vumrao opened the pull request:
Signed-off-by: Vikhyat Umrao <vumrao@redhat.com>

vumrao commented:
http://tracker.ceph.com/issues/4243

vumrao commented:
$./rbd create -p rbd test -s 100M --image-format 2
$./rbd -p rbd ls -l
NAME    SIZE PARENT FMT PROT LOCK 
test 102400k          2           
$ ./rbd snap create test@testsnap
$ ./rbd -p rbd ls -l
NAME             SIZE PARENT FMT PROT LOCK 
test          102400k          2           
test@testsnap 102400k          2           

 image@snap syntax works fine !

dillaman commented on src/rbd.cc:
```diff
@@ -3391,13 +3401,6 @@ if (!set_conf_param(v, p1, p2, p3)) { \
 	 << std::endl;
     return EXIT_FAILURE;
   }
-  if ((opt_cmd == OPT_SNAP_CREATE || opt_cmd == OPT_SNAP_ROLLBACK ||
```
Why remove this single check just so you can re-add it for each individual command below?

dillaman commented on src/rbd.cc:
```diff
@@ -3586,6 +3589,10 @@ if (!set_conf_param(v, p1, p2, p3)) { \
     break;
 
   case OPT_CLONE:
+    if (!imgname || !snapname) {
```
Now that you check for an imgname starting at line 3330, these individual checks for a populated imgname can be removed from each command block.

dillaman commented on src/rbd.cc:
```diff
@@ -100,9 +100,9 @@ void usage()
-"  children <snap-name>                        display children of snapshot\n"
+"        --snap <parentsnap> <image-name> <clonename>\n"
+"                                              clone a snapshot into a COW child image\n"
+"  children --snap <snap-name> <image-name>    display children of snapshot\n"
```
Since ```snap-name``` is documented as ```[pool/]name[@snap]``` in the help, perhaps we should use just ```snap``` when the intent is only a snapshot name within an image.

jdurgin commented:
I agree with Jason's comments. If you want to say 'rbd: ' command 'requires snap and image name' we can do that in one place, rather than in each command. A helper to translate OPT_* into strings would be useful for that.

vumrao commented:
Thanks Jason and Josh for your comments, sure I will modify the patch.

vumrao commented:
I have reverted my multiple places check for image and snap name and also older code which checks it.

As this code part takes care of snap-name check part -

 if ((opt_cmd == OPT_SNAP_CREATE || opt_cmd == OPT_SNAP_ROLLBACK ||
       opt_cmd == OPT_SNAP_REMOVE || opt_cmd == OPT_CLONE ||
       opt_cmd == OPT_SNAP_PROTECT || opt_cmd == OPT_SNAP_UNPROTECT ||
       opt_cmd == OPT_CHILDREN) && !snapname) {
    cerr << "rbd: snap name was not specified" << std::endl;
    return EXIT_FAILURE;
  }  

and below given code takes care of imagename part 

if (opt_cmd != OPT_LIST &&
      opt_cmd != OPT_IMPORT &&
      opt_cmd != OPT_IMPORT_DIFF &&
      opt_cmd != OPT_UNMAP && /* needs imgname but handled below */
      opt_cmd != OPT_SHOWMAPPED &&
      opt_cmd != OPT_MERGE_DIFF &&
      opt_cmd != OPT_DISK_USAGE && !imgname) {
    cerr << "rbd: image name was not specified" << std::endl;
    return EXIT_FAILURE;
  }

@jdurgin , we can add command name in error message but i think it is not needed as this error message would be for same command in stdout.

@dillaman , I had changed snap-name to snap , I hope this is what you meant , please let me know if I misunderstood it.

Thanks and Regards,
Vikhyat

vumrao commented:
@dillaman , sorry I think I cherry-picked it and then submitted once again I think some mistake I did while recommitting it , will resubmit it.
Thanks,
Vikhyat

jdurgin commented:
Looks good to me. @dillaman what do you think?

dillaman commented:
:+1:

idryomov commented:
```
What I think we should be after here is stressing the fact that image-name is
in fact [pool]/image, snap-name is in fact [pool]/image@snap and disambiguating
it from cases where e.g. snap-name is just the name, like rbd diff --from-snap
option argument.  I propose changing those image-name instances to image-spec
and those snap-name instances to snap-spec, so that the help reads:

clone ... <snap-spec> <image-spec>
children <snap-spec>
...
diff [--from-snap <snap-name>] ... <image-spec>
...
snap create <snap-spec>
...
<image-spec> is [pool-name]/image-name,
<snap-spec> is [pool-name]/image-name@[snap-name],
or you may specify individual pieces with -p/--pool <pool-name>,
--image <image-name> and --snap <snap-name>.

Having "children --snap <snap> <image-name>" is not nice at the very least, as
--snap is not required there and IMO we should be pushing people towards the
spec syntax.  (In fact, I'd be in favor of deprecating --pool, --image, --snap,
--dest-pool and --dest entirely - specs are much easier to read.  Compare

    $ rbd clone --pool a --image b --snap c --dest-pool d --dest e

with

    $ rbd clone a/b@c d/e

Unfortunately our docs are all littered with --pool and --snap so a lot of
people don't even know about this syntax.)
```

idryomov commented:
Also, you should change the title so that it's clear that the main objective of the patch is to fix help and then clarify in the commit message why you removed those checks - I had to go and look to make sure they were redundant.  (BTW, I think you can remove OPT_EXPORT && !imgname check as well while you are at it.)

jdurgin commented:
I like that approach @idryomov - it's much clearer with the snap-spec and image-spec rather than names, and I agree the spec syntax is much easier to read. We should update the docs to use it more consistently.

vumrao commented:
@jdurgin yes same here I also liked the approach will make the changes to the patch and will send another pull request for doc changes.

Thanks @idryomov  for suggesting us a nice approach.

vumrao commented:
@jdurgin  @idryomov  I have sent two patches one for code changes and one for doc changes. 

Regards,
Vikhyat

idryomov commented on src/rbd.cc:
```diff
@@ -95,71 +95,71 @@ void usage()
+"<image-spec> is [pool-name]/image-name and <snap-spec> is [pool-name]/image-name@snap-name\n"
+"or you may specify individual pieces of names with -p/--pool <pool-name>, --image <image-name>\n"
+"and/or --snap <snap-name>.\n"
 "\n"
```
This is too long, split it the way I had it split in my comment.

idryomov commented:
Looks good, but there are a few minor issues that are easier to fix than to comment on.  I'll do a manual merge tomorrow.  Thanks!

vumrao commented:
Thanks @idryomov !

idryomov commented:
Superseded by #5352, closing.
